ddl/ddl_api.go

```diff
@@ -838,50 +838,22 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 	if err != nil {
 		return errors.Trace(err)
 	}
-	if s.Partition != nil {
-		pi := &model.PartitionInfo{
-			Type:   s.Partition.Tp,
-			Enable: ctx.GetSessionVars().EnableTablePartition,
-		}
-		if s.Partition.Expr != nil {
-			buf := new(bytes.Buffer)
-			s.Partition.Expr.Format(buf)
-			pi.Expr = buf.String()
-			if s.Partition.Tp == model.PartitionTypeRange {
-				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
-					for _, col := range cols {
-						name := strings.Replace(col.Name.String(), ".", "`.`", -1)
-						if !(col.Tp == mysql.TypeLong || col.Tp == mysql.TypeLonglong) && fmt.Sprintf("`%s`", name) == pi.Expr {
-							return errors.Trace(ErrNotAllowedTypeInPartition.GenByArgs(pi.Expr))
-						}
-					}
-				}
-			}
-		} else if s.Partition.ColumnNames != nil {
-			pi.Columns = make([]model.CIStr, 0, len(s.Partition.ColumnNames))
-			for _, cn := range s.Partition.ColumnNames {
-				pi.Columns = append(pi.Columns, cn.Name)
-			}
+
+	part, err := buildTablePartitionInfo(ctx, d, s, cols)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if part != nil && part.Type == model.PartitionTypeRange {
+		err = checkPartitionNameUnique(tbInfo, part)
+		if err != nil {
+			return errors.Trace(err)
 		}
-		for _, def := range s.Partition.Definitions {
-			// TODO: generate multiple global ID for paritions.
-			pid, err1 := d.genGlobalID()
-			if err1 != nil {
-				return errors.Trace(err1)
-			}
-			piDef := model.PartitionDefinition{
-				Name:    def.Name,
-				ID:      pid,
-				Comment: def.Comment,
-			}
-			for _, expr := range def.LessThan {
-				buf := new(bytes.Buffer)
-				expr.Format(buf)
-				piDef.LessThan = append(piDef.LessThan, buf.String())
-			}
-			pi.Definitions = append(pi.Definitions, piDef)
+		err = checkCreatePartitionValue(part)
+		if err != nil {
+			return errors.Trace(err)
 		}
-		tbInfo.Partition = pi
+		tbInfo.Partition = part
```

Comment:
If we only get the partition of type "PartitionTypeRange"，why do we check it so many places？